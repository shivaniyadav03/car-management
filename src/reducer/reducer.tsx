
const InitialValue={
    numberOfLots:[{
        lotid:1,
        assignedTo: null
    },
    {
        lotid:2,
        assignedTo: null
    },
    {
        lotid:3,
        assignedTo: null
    },
    {
        lotid:4,
        assignedTo: null
    },
    {
        lotid:5,
        assignedTo: null
    }
]
};

const randomIntFromInterval = (min: number, max: number) => { 
    return Math.floor(Math.random() * (max - min + 1) + min)
}

 const reducer=(state:any=InitialValue,action:any)=>{
 switch(action.type){
    case 'ASSIGN_LOT':{
        const currentNumberOfLots: any[] = state?.numberOfLots
        const unAssignedLots:any[] = currentNumberOfLots.filter((item:any)=>{
            return item.assignedTo==null;
        });
        Object.keys(action.payload).map(item => {
            const storeInRandomIndex = () => {
                const randomIndex: number = randomIntFromInterval(0,unAssignedLots.length-1);
                if(unAssignedLots[randomIndex]?.assignedTo){
                    storeInRandomIndex()
                }else{
                    unAssignedLots[randomIndex].assignedTo = action.payload[item];
                    alert('Assign Lots:'+randomIndex);
                }
            }
            storeInRandomIndex();
        })
        unAssignedLots.map((item: any) => {
            const currentIndex: number = currentNumberOfLots.findIndex((x: any) => item.lotid == x.lotid);
            if(currentIndex !== -1){
                currentNumberOfLots[currentIndex].assignedTo = item.assignedTo
            }
        })
        return {
            ...state,
            numberOfLots: currentNumberOfLots
        }
    }
    case 'EXIT_LOT':{
        const currentNumberOfLots:any[]=state?.numberOfLots;
         const updatedLots: any[] = currentNumberOfLots.map((item:any,index:number)=>{
          if(item.lotid==action.payload){
            item.assignedTo=null;            
          }
          return item;
         });
         return{
             ...state,
             numberOfLots: updatedLots
         }
    }
    default:return {
        ...state
    }
 }
}
export default reducer;