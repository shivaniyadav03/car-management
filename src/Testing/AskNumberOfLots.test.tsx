import {cleanup, findByTestId, fireEvent, render} from '@testing-library/react';
import { Provider } from 'react-redux';
import AskNumberOfLots from '../Pages/askNumberOfLots';
import {createStore} from 'redux';
import reducer from '../reducer/reducer';
//import {useSelector} from 'react-redux';

beforeEach(()=>{
    document.body.innerHTML="";
});
afterEach(()=>{
    cleanup();
});
const mockedData=jest.fn();
jest.mock('react-router-dom',()=>{
    const originalDataDom=jest.requireActual('react-router-dom');
    return{
        __esModule:true,
        ...originalDataDom,
        useNavigate:()=>mockedData,
        useLocation:()=>mockedData,
        useSelector:()=>mockedData
    };
});

describe('Write Test Case for AskNumberOfLots Component',()=>{
    const store=createStore(reducer);
    render(
        <Provider store={store}>
            <AskNumberOfLots/>
        </Provider>
    );
    test('To create snapshots of AskNumberOfLots Component',()=>{
        expect(screen).toMatchSnapshot();
        });
    test('Test Case For InitialState Of Reducer',()=>{
    expect(reducer(undefined,{})).toEqual({
        numberOfLots:[{
            lotid:1,
            assignedTo: null
        },
        {
            lotid:2,
            assignedTo: null
        },
        {
            lotid:3,
            assignedTo: null
        },
        {
            lotid:4,
            assignedTo: null
        },
        {
            lotid:5,
            assignedTo: null
        }
    ]
    
    });
});

// test('Test case for Button',()=>{
//     const buttonId=getByRole('button',{name:/Ask For No Of Lots Available/i});
// });

});