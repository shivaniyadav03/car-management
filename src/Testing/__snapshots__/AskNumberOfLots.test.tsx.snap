// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Write Test Case for AskNumberOfLots Component To create snapshots of AskNumberOfLots Component 1`] = `Screen {}`;
